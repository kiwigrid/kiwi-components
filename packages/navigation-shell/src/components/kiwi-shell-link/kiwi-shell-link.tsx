import { Component, h, Prop } from '@stencil/core';
import {
  isActive,
  makeLink,
} from '../kiwi-navigation-shell/kiwi-navigation-shell.store';

@Component({
  tag: 'kiwi-shell-link',
  shadow: false,
})
export class KiwiShellLink {
  /** The key of the route config this link should be built off of. */
  @Prop()
  public routeKey!: string;

  /** Data associated to this route. */
  @Prop()
  public routeData: Record<string, unknown> = {};

  /** Additional css to be applied to the underlying `a` element. */
  @Prop()
  public customClass?: string;

  /**
   * If set to true will append class 'active' to the `a` element. Alternatively
   * provide a string which is used as active class instead.
   * When set to false, no class will be attached (default).
   */
  @Prop()
  public activeClass: boolean | string = false;

  render(): JSX.Element {
    try {
      const [url, label, handler] = makeLink(this.routeKey, this.routeData);

      return (
        <a href={url} onClick={handler} class={this.classMap}>
          <slot>{label}</slot>
        </a>
      );
    } catch (e) {
      console.warn(e);

      return <slot>{this.routeKey}</slot>;
    }
  }

  private get classMap(): Record<string, boolean> {
    const activeClass =
      this.activeClass === true ? 'active' : `${this.activeClass}`;

    return {
      [this.customClass ?? '']: typeof this.customClass === 'string',
      [activeClass]: this.activeClass !== false && isActive(this.routeKey),
    };
  }
}
