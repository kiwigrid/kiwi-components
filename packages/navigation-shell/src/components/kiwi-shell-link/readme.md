# kiwi-link



<!-- Auto Generated Below -->


## Properties

| Property                | Attribute      | Description                                                                                                                                                                                 | Type                        | Default     |
| ----------------------- | -------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------- | ----------- |
| `activeClass`           | `active-class` | If set to true will append class 'active' to the `a` element. Alternatively provide a string which is used as active class instead. When set to false, no class will be attached (default). | `boolean \| string`         | `false`     |
| `customClass`           | `custom-class` | Additional css to be applied to the underlying `a` element.                                                                                                                                 | `string \| undefined`       | `undefined` |
| `routeData`             | --             | Data associated to this route.                                                                                                                                                              | `{ [x: string]: unknown; }` | `{}`        |
| `routeKey` _(required)_ | `route-key`    | The key of the route config this link should be built off of.                                                                                                                               | `string`                    | `undefined` |


## Dependencies

### Used by

 - [kiwi-shell-breadcrumb](../kiwi-shell-breadcrumb)

### Graph
```mermaid
graph TD;
  kiwi-shell-breadcrumb --> kiwi-shell-link
  style kiwi-shell-link fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
