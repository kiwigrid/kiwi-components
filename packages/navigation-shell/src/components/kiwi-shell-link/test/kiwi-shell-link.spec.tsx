import { newSpecPage } from '@stencil/core/testing';
import { KiwiShellLink } from '../kiwi-shell-link';

describe('kiwi-shell-link', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [KiwiShellLink],
      html: `<kiwi-shell-link></kiwi-shell-link>`,
    });
    expect(page.root).toEqualHtml(`
      <kiwi-shell-link>
        <mock:shadow-root>
          <slot></slot>
        </mock:shadow-root>
      </kiwi-shell-link>
    `);
  });
});
