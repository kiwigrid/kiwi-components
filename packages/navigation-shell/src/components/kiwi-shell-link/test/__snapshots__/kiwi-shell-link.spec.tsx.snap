// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kiwi-shell-link renders 1`] = `
<kiwi-shell-link>
  <!---->
  <a href="/route/66">
    <slot-fb>
      Route 66
    </slot-fb>
  </a>
</kiwi-shell-link>
`;

exports[`kiwi-shell-link renders custom label 1`] = `
<kiwi-shell-link>
  <!---->
  <a href="/route/66">
    <span class="my-icon"></span>
    <slot-fb hidden="">
      Route 66
    </slot-fb>
  </a>
</kiwi-shell-link>
`;

exports[`kiwi-shell-link renders dynamic label 1`] = `
<kiwi-shell-link>
  <!---->
  <a href="/users/johndoe-42">
    <slot-fb>
      User johndoe-42
    </slot-fb>
  </a>
</kiwi-shell-link>
`;
